README.md
# Homebridge Plugin for OpenHAB2 - Complete Edition (works also with OpenHAB3!)

[![NPM](https://nodei.co/npm/homebridge-openhab2-complete.png)](https://nodei.co/npm/homebridge-openhab2-complete/)

This [homebridge](https://github.com/nfarina/homebridge) plugin for [openHAB](https://www.openhab.org) fully supports all services offered by Apple's HomeKit Accessory Protocol (HAP), as far as it is feasible based on the item types offered by OpenHAB (see [below](#supported-hap-services) for the currently supported 32 accessories). In opposite to the existing [openHAB homebridge plugin](https://www.npmjs.com/package/homebridge-openhab2) or the native [openHAB Homekit Plugin](https://www.openhab.org/addons/integrations/homekit/), this plugin requires explicit declaration of accessories in the homebridge configuration and does not use openHAB's tagging system, which leads to a little more effort during configuration, but proves more reliable and functional in more complex installations. See [Comparisson](#comparison) below.

## Preface
~~Since the native HomeKit integration of OpenHAB has significantly improved (still lacking some of the available Service of this plugin), I don't think there is much value in this plugin.~~

After testing openHAB3's HomeKit integration for a week, I decided to move back to homebridge, due to the instability of openHAB's implementation.

As part of my OpenHAB 3 migration I  updated the code to support OH3, finally put some bug fixes that happened over the last year into the release and replaced the depricated `request` package. 

To pull the Github version:
```
npm install -g -f steilerDev/homebridge-openhab2-complete#master
```

## Installation
*Note: Please install [homebridge](https://www.npmjs.com/package/homebridge) first.*

```
npm install -g homebridge-openhab2-complete
```

Some people are experiencing dependency issues between homebridge's node version and the required node version for this project. My local setup is based on [oznu's homebridge docker container](https://github.com/oznu/docker-homebridge), where I never ran into any problems. In order to install the plugin in the docker, just add `npm install homebridge-openhab2-complete` to the `startup.sh` script inside the mapped docker volume.

## Configuration
This is a platform plugin, that will register all accessories within the Bridge provided by homebridge. The following shows the general homebridge configuration (`config.json`), see the [Supported HAP Services below](#supported-hap-services), in order to get the detailed configuration for each service.

```json
{
    "bridge": {
        ...
    },

    "accessories": [
        ...
    ],

    "platforms": [
        {
            "platform": "openHAB2-Complete",
            "host": "http://192.168.0.100",
            "port": "8080",
            "username": "homebridge",
            "password": "homebridge",
            "accessories": [
                {
                    "name": "An items name, as shown in Homekit later",
                    "type": "switch",
                    "item": "Itemname-within-OpenHAB"
                },
                ...
            ]
        },
        ...
    ]
}
```
* `platform` has to be `"openHAB2-Complete"`
* `host`: The IP or hostname of your openHAB instance. The Protocol specifier (`http://`) is optional, defaults to `http://` (independent of the specified port)
* `port`: *(optional)* If not specified the default port of the specified `host` protocol is used
* `username`: *(optional)* Username for HTTP(S) basic auth
* `password`: *(optional)* Password for HTTP(S) basic auth
* `accessory`: An array of accessories exposed to HomeKit, see the next chapter for available services and their configurations

Alternatively you can group accessories of the same `type` in a sub-array:
```json
...
    "platforms": [
        {
            ...
            "accessories": [
                {
                    "type": "switch",
                    "items": [
                        {
                            "name": "An items name, as shown in Homekit later",
                            "item": "Itemname-within-OpenHAB"
                        },
                        {
                            "name": "An items name, as shown in Homekit later",
                            "item": "Itemname-within-OpenHAB"
                        },
                        ...
                    ]
                },
                ...
            ]
        },
...
```

## Supported HAP Services
The following is a list of all services that are currently supported and which values are required within the accessory configuration. Every accessory needs a `name` (as shown in HomeKit later) and a `type`. The combination of `name` and `type` needs to be unique and should not change in the hombridge configuration (you can safely rename the item within the Home.app after adding them). Changing `name` or `type` in the config will change the item's serial number, which would be registered as a new/different one.

* Complex Accessories:
  * [Lightbulb](#lightbulb) 
    * Homebridge configuration type: `light`
  * [Fan](#fan)
    * Homebridge configuration type: `fan`
  * [Security System](#security-system)
    * Homebridge configuration type: `security`
* Climate Control Accessories:
  * [Thermostat](#thermostat)
    * Homebridge configuration type: `thermostat`
  * [Humidifier/Dehumidifier](#humidifierdehumidifier)
    * Homebridge configuration type: `humidifier`
  * [Heater/Cooler](#heatercooler)
    * Homebridge configuration type: `heatercooler`
  * [Air Purifier](#air-purifier)
    * Homebridge configuration type: `airpurifier`
* Audio Accessories:
  * [Speaker](#speaker)
    * Homebridge configuration type: `speaker`
  * [Microphone](#microphone)
    * Homebridge configuration type: `microphone`
* Watering Accessories:
  * [Valve](#valve), which can be an 'Irrigation System' or 'Faucet'
    * Homebridge configuration type: `valve`
* Position Based Actors:
  * [Window Covering](#window-covering)
    * Homebridge configuration type: `windowcovering`
  * [Door](#door)
    * Homebridge configuration type: `door`
  * [Window](#window)
    * Homebridge configuration type: `window`
  * [Lock Mechanism](#lock-mechanism)
    * Homebridge configuration type: `lock`
  * [Garage Door Opener](#garage-door-opener)
    * Homebridge configuration type: `garage`
  * [Slat](#slat)
    * Homebridge configuration type: `slat`
* Numeric Sensors:
  * [Temperature Sensor](#temperature-sensor)
    * Homebridge configuration type: `temp`
  * [Humidity Sensor](#humidity-sensor)
    * Homebridge configuration type: `humidity`
  * [Light Sensor](#light-sensor)
    * Homebridge configuration type: `lux`
  * [Air Quality Sensor](#air-quality-sensor)
    * Homebridge configuration type: `air`
* Binary Actors:
  * [Switch](#switch)
    * Homebridge configuration type: `switch`
  * [Outlet](#outlet)
    * Homebridge configuration type: `outlet`
* Binary Sensors:
  * [Motion Sensor](#motion-sensor)
    * Homebridge configuration type: `motion`
  * [Occupancy Sensor](#occupancy-sensor)
    * Homebridge configuration type: `occupancy`
  * [Leak Sensor](#leak-sensor)
    * Homebridge configuration type: `leak`
  * [Carbon Monoxide Sensor](#carbon-monoxide-sensor)
    * Homebridge configuration type: `co`
  * [Carbon Dioxide Sensor](#carbon-dioxide-sensor)
    * Homebridge configuration type: `co2`
  * [Contact Sensor](#contact-sensor)
    * Homebridge configuration type: `contact`
  * [Smoke Sensor](#smoke-sensor)
    * Homebridge configuration type: `smoke`
  * [Filter Maintenance Sensor](#filter-maintenance-sensor)
    * Homebridge configuration type: `filter`
  
The following services are also defined by the HomeKit protocol, but since I don't know a good way to map them to openHAB items, I currently don't plan to implement them. Let me know if you have any ideas, by opening an issue!
* Camera RTP Stream Management
* Doorbell
* Stateless Programmable Switch

## Configuration for every accessory
### Grouping
Since iOS 13 multiple accessories can be grouped within a single accessory. This can be -as far as I have tested- done in an arbitrary way. The syntax to define a grouped item is as follows:
```json
{
    {
        "type": "group",
        "name": "Group Name",
        "model": "Group Type Model",
        "batteryItem": "Itemname-within-OpenHAB",
        "batteryItemThreshold": "20",
        "items": [
            {
                "type": "homebridge-openhab2-complete item type",
                "name": "An items name, as shown in Homekit later",
                "item": "Itemname-within-OpenHAB"
            }, 
            {
                "type": "homebridge-openhab2-complete item type",
                "items": [
                    {
                        "name": "An items name, as shown in Homekit later",
                        "item": "Itemname-within-OpenHAB",
                    },
                    {
                        ...
                    }
                ]
            }
        ]
    }
}
```
* `type`: Needs to be `"group"`
* `name`: The name as shown in HomeKit later (in some instances I have experienced that this name is dropped in favor of the first item's name)
* `model` *(optional)*: A model description, defining the grouping. This is only shown in the expanded settings of the group.
* `batteryItem`, `batteryItemThreshold`, ... *(optional)*: A battery service can only be defined on the most upper level for the whole group. See [Battery Levels and Warnings](#battery-levels-and-warnings) for more information on the syntax. All `BatteryServices` defined in the `items` array will be ignored!
* `items`: An array of item definitions. As shown in the example, the different items will be defined here, as if they were outside of the group, either through single item definition or multiple definitions of the same type in a sub-array. Those configuration styles can be combined!

### Battery Levels and Warnings
Every accessory can be configured to show battery warnings and battery levels. The following configuration can be optionally added to every item:
```json
{
    "batteryItem": "Itemname-within-OpenHAB",
    "batteryItemThreshold": "10",
    "batteryItemInverted": "false",
    "batteryItemChargingState": "Itemname-within-OpenHAB",
    "batteryItemChargingStateInverted": "false"
}
```

* `batteryItem` *(optional)*: The openHAB item representing the battery level for this accessory.
  * Needs to be of type `Switch`, `Contact` or `Number` within openHAB
  * **Note**: If the type is `Number`, then the battery warning will be based on `batteryItemThreshold` and the battery level will be the value of the item
  * **Note**: If the type is `Switch` or `Contact` the `BatteryLevel` will always show 0% and for states `ON` or `OPEN` a battery warning will be triggered
* `batteryItemInverted` *(optional)*: If `batteryItem`'s state needs to be interpreted inverted, set this value to `"true"` (Only for types `Switch` and `Contact`)
  * Default: `"false"`
  * Allowed values: `"true"` & `"false"` *don't forget the quotes*
* `batteryItemThreshold` *(optional)*: The threshold of `batteryItem`'s numeric value, below a warning will be triggered
  * Default: `"20"`
  * Allowed values: All integers
* `batteryItemChargingState`: The openHAB item representing if the device is currently charging
  * Needs to be of type `Switch` or `Contact` within openHAB
  * Default: Item will be reported as 'Not chargeable' to HomeKit if this item is not available
* `batteryItemChargingStateInverted` *(optional)*: If `batteryItemChargingState`'s state needs to be interpreted inverted, set this value to `"true"`
  * Default: `"false"`
  * Allowed values: `"true"` & `"false"` *don't forget the quotes*
  
## Available accessory types
### Lightbulb
This service describes a lightbulb.

```json
{
    "name": "An items name, as shown in Homekit later",
    "type": "light",
    "item": "Itemname-within-OpenHAB",
    "sendOnOnlyWhenOff": "false"
}
```
* `item`: The openHAB item controlled by this accessory
  * Needs to be of type `Switch`, `Dimmer` or `Color` within openHAB (HomeKit will correctly display brightness *-in case of `Dimmer` or `Color`-* and color settings *-in case of `Color`-*)
* `sendOnOnlyWhenOff`: This addresses the different behaviour of lights as discussed in issue #57 and when enabled uses @weakfl 's proposed fix
  * Default: `"false"`
  * Allowed values: `"true"` & `"false"` *don't forget the quotes*

  
### Fan
This service describes a fan.

```json
{
    "name": "An items name, as shown in Homekit later",
    "type": "fan",
    "item": "Itemname-within-OpenHAB"
}
```
* `item`: The openHAB item controlled by this accessory
  * Needs to be of type `Switch`, `Number` or `Dimmer` within openHAB (HomeKit will correctly display fan speed control *-in case of `Number` or `Dimmer`-*)

### Security System
This service describes a security system.

```json
{
    "name": "An items name, as shown in Homekit later",
    "type": "security",
    "item": "Itemname-within-OpenHAB"
}
```
* `item`: The openHAB item representing the state of the security system
  * Needs to be of type `String` within openHAB
  * Allowed values (Every other value will be ignored and shown as an error)
    * `StayArm`: The home is occupied and the residents are active, shown as `Home` in the Home.app
    * `AwayArm`: The home is unoccupied, shown as `Away` in the Home.app
    * `NightArm`: The home is occupied and the residents are sleeping, shown as `Night` in the Home.app
    * `Disarmed`: The security system is disabled, shown as `Off` in the Home.app
    * `AlarmTriggered`: The security alarm is triggered

### Thermostat
This service describes a thermostat.

*Important notes on Thermostat Capabilities*
* A thermostat can have the capability to cool, heat or do both
* The `modeItem` will precede `heatingItem` or `coolingItem` (when `modeItem` is defined, `heatingItem` and `coolingItem` will be ignored)
  * If `modeItem` is present while `modeItemCharacteristic` is present the allowed values for `modeItem` will be restricted based on `modeItemCharacteristic` (see below)
* If `modeItem` is not present, one of the following is required: `heatingItem` or `coolingItem`
  * Based on those, the capability to heat or cool will be derived
  * If both are present, `coolingItem` will be ignored, therefore the item will only be able to either cool or heat

*Important notes on Thermostat Capabilities*

```json
{
    "name": "An items name, as shown in Homekit later",
    "type": "thermostat",
    "currentTempItem": "Itemname-within-OpenHAB",
    "targetTempItem": "Itemname-within-OpenHAB",
    "currentHumidityItem": "Itemname-within-OpenHAB",
    "targetHumidityItem": "Itemname-within-OpenHAB",
    "heatingItem": "Itemname-within-OpenHAB",
    "coolingItem": "Itemname-within-OpenHAB",
    "tempUnit": "Celsius",
    "heatingThresholdTempItem": "Itemname-within-OpenHAB",
    "coolingThresholdTempItem": "Itemname-within-OpenHAB",
    "modeItem": "Itemname-within-OpenHAB",
    "modeItemCapability": "HeatingCooling",
    "minTemp": "-100",
    "maxTemp": "200",
    "minTempStep": "0.1"
}
```


* `currentTempItem`: The openHAB item representing the current temperature as measured by the thermostat
  * Needs to be of type `Number`, `Rollershutter`, or `Dimmer` within openHAB 
  
    *(Note: When using `Dimmer` or `Rollershutter` type and OpenHAB receives a non numeric command like `ON`, `OFF`, `INCREASE`, `DECREASE`, `UP` or `DOWN` this might lead to unexpected behaviour and/or non-responsive HomeKit items. This exception is not covered by this plugin and the user needs to ensure a consistent state)*
* `targetTempItem`: The openHAB item representing the target temperature inside the room

CHANGELOG.md
# Changelog of homebridge-openhab2-complete
## Changelog

### V1.3.6 (unreleased)
* Hide unnecessary Heater Cooler modes (Thanks @Matsuo3rd)
* Fix minTemp & maxTemp scaling (Thanks @Matsuo3rd)
* Misc. fixes on humidifier / dehumidifier (Thanks @Matsuo3rd)

### V1.3.5
* Removed `Faucet` and `Irrigation System` in favor of generic `Valve`
* Added configuration options for HTTP(S) basic authentication
  * Making basic auth optional (Thanks @Matsuo3rd)
* Swing item now read-write
* Fixed swing item usage in Binary.js (Thanks @Matsuo3rd)
* Complying with openHAB's correct event subscription filtering (see https://github.com/openhab/openhab-core/pull/2986, thanks @Matsuo3rd)

#### Issues
* Closes #105

### V1.3.4
* Added option to change light behaviour with `sendOnOnlyWhenOff` option
* Fixed documentation

#### Issues
* Closes #57
* Closes #102

### V1.3.3
* Fixed wrong variable name

#### Issues
* Closes #100

### V1.3.2
  * Missed one occurence of the `request` dependency

#### Issues
* Closes #99

### V1.3.0
#### Important Information
This is proabably the last release of this plugin. Since the native HomeKit integration of OpenHAB has significantly improved (still lacking some of the available Service of this plugin), I don't think there is much value in this plugin. As part of my OpenHAB 3 migration I  updated this code to support OH3, finally put some bug fixes that happened over the last year into a release and replaced the depricated `request` package. So I am certain this plugin will continue functioning, and I will continue merging PRs into upstream, but don't expect a new release, unless my test run of OH3's native HomeKit integration fails.

#### Breaking changes
I reworked the UUID generation, now using the included mechanism (Issue #88). This however will probably reset the items already added to Home.app (once!). Please be aware that there will probably be some work required to re-organize the Home.app!

* OpenHAB 3 Support (Thanks @D-J-See and @DanielKnoop)
* Fixed UUID generation issue by using supplied function
* Fixed `TypeError: service.on is not a function` error during item creation
* Honoring homebridge's item caching
* Exchanged depricated `request` for `needle`

#### Issues
* Closes #88

### V1.2.0
* Extended `NumericSensor` and `NumericSensorActor` capabilities to support `Rollershutter` and `Dimmer` types.
* Homebridge no longer crashes if openHAB's host is not available
* Support for scientifically notated numbers (e.g. `7E+1`)
* Fixed wrong heater/cooler state
* Enabling support for user-defined min/max values:
  * minTemp, maxTemp, minTempStep for Thermostat, HeaterCooler and Temperature Sensor
  * minFanSpeed, maxFanSpeed, minFanStep for HeaterCooler, AirPurifier and Humidifier/Dehumidifier
* Fixed documentation issue around `activeItem` and `activeItemInverted` for Air Purifier, Irrigation System, Humidifier/Dehumidifier.
* Fixed missing `LEVEL_CONFIG` for Filter Maintenance Sensor, CO Sensor and CO2 Sensor
* Reworked thermostat and added warning for Heater/Cooler, see README
* Added `durationItemMax` to the Valve Service

#### Issues:
* Closes #60
* Closes #70
* Closes #73
* Closes #74
* Closes #79
* Closes #80
* Closes #82
* Closes #83
* Closes #84

### V1.1.0
* Updated Readme to fix typo's
* Added possibility to arbitrary group items using the grouping feature of iOS 13
* Added slat type accessory
* Added audio type accessories: Microphone & Speaker

### V1.0.0
* Updated Readme to fix typo's
* Valve and Faucet should now be able to change state of openHAB item (changed configuration and requirements for Valve and Faucet items, see README)
* 'heatingItem' and 'coolingItem' of Thermostat now actually support 'Contact' type
* Reworked 'Thermostat' and 'Temperature Sensor' to support a wider range of numbers (now from -100 to +200) and also support exotic temperature units (Fahrenheit :P)
* Reworked light based on issues #13 and #43
* Added numeric battery state and battery warnings for _every_ accessory type
* Added vertical and horizontal tilt angle's for `Window Covering`'s slats
* Reworked `Security System` behaviour, using a single `String` item in comparision to a complex `Switch` construct

#### Issues:
* Closes #9
* (Hopefully finally) closes #13
* Closes #17
* Closes #43
* Closes #24
* Closes #33
* Closes #38
* Closes #32
* Closes #34
* Closes #35

**Breaking changes**:
* Changed `item` to `currentTempItem` within `temp` accessory
* `Valve` and `Faucet` configuration and requirements changed significantly, consult the relevant section in the `README.md`
* `Security System` moved to a single `String` item within OpenHAB

### V0.10.2
* Fixed minor issues from Github.

### V0.10.1
* Fixed a logic bug for inverted `Window Covering` Accessory using inverted `Rollershutter` items
* Fixed light off bug

#### Issues:
* Closes #11, #12, #13

### V0.10.0
* Allowing grouping of same type accessory inside configuration
* Added Air Quality Sensor
* Added Climate Control Accessories:
  * Humidifier/Dehumidifier
  * Heater/Cooler
  * Air Purifier
* Added Watering Accessories:
  * Faucet
  * Valve
  * Irrigation
* Added Garage Door Opener Accessory
* Support for `Number` types with Unit of Measurement
  
**Breaking changes**:
* Renamed light sensor's accessory type from `lightSensor` to `lux`

#### Issues:
* Closes #8

### V0.9.0
* Removed `clone` dependency, since it caused unexpected behaviour
* Improved fan in order to support non-binary items to control the fan speed
* Fixed an issue, where not supported characteristics were still being exposed to HomeKit
* Reworked thermostat to extend usage for AUTO mode and mode item

#### Issues:
* Closes #2, #4, #5, #7

### V0.8.2
* Created own caching mechanism, supporting this use case & improving UX performance & stress to openHAB.
* Improved startup performance due to bulk sync of item states

### V0.8.0

util/OpenHAB.js
'use strict';

const {URL} = require('url');
const syncRequest = require('sync-request');
const EventSource = require('eventsource');
const {Cache} = require('./Cache');
var needle = require('needle');

// 30 mins ttl for cached item states
const valueCacheTTL = 30 * 60 * 1000;
// Checking every minute, if item states from the cache need to be cleared
const monitorInterval = 5 * 60 * 1000;

class OpenHAB {

    constructor(config, log) {
        this._log = log;

        if(config.host.startsWith("http://") || config.host.startsWith("https://")) {
            this._hostname = config.host;
        } else {
            this._hostname = `http://${config.host}`;
        }
        this._port = config.port;
        this._username = config.username;
        this._password = config.password;

        this._valueCache = new Cache(log, valueCacheTTL, monitorInterval);

        this._valueCache.on('expired', function (habItem) {
            this._log.warn(`Item ${habItem}'s state was cleared from the cache, getting the current value`);
            this._getStateWithoutCache(habItem, function(error, value) {
                if(error) {
                   this._log.error(`Unable to set new value for ${habItem}: ${error.message}`);
                } else {
                    if(this._subscriptions[habItem] !== undefined) {
                        this._log.debug(`Received new state for item ${habItem}: ${value}`);
                        this._subscriptions[habItem].forEach(function (callback) {
                            callback(value, habItem);
                        });
                    } else {
                        this._log.debug(`Not pushing new value for ${habItem}, because it is not registered with homebridge`);
                    }
                }
            }.bind(this));
        }.bind(this));

        this._typeCache = new Cache(log);

        this._subscriptions = {};
    }

    _getURL(pathname, search) {
        let newURL = new URL(this._hostname);
        if(this._port !== undefined) {
            newURL.port = this._port;
        }
        if(pathname) {
            newURL.pathname = pathname;
        }
        if(search) {
            newURL.search = search;
        }
        let url = {};
        url.public = newURL.href;
        if (this._username && this._password) {
            newURL.username = this._username;
            newURL.password = this._password;
        }
        url.href = newURL.href;
        return url;
    }

    isOnline() {
        let myURL = this._getURL(`/rest/items`);
        try {
          const response = syncRequest('GET', myURL);
          this._log.debug(`Online request for openHAB (${myURL.public}) resulted in status code ${response.statusCode}`);
          return response.statusCode === 200;
        } catch (e) {
          this._log.warn(`Unable to retrieve openHAB URL ${myURL.public}: ${e}`);
          return false;
        }
    }

    getState(habItem, callback) {
        let cachedValue = this._valueCache.get(habItem);
        if(cachedValue) {
            this._log.debug(`Getting value for ${habItem} from the cache`);
            callback(null, cachedValue);
        } else {
            this._log.warn(`Getting value for ${habItem} from openHAB, because no cached state exists`);
            this._getStateWithoutCache(habItem, callback);
        }
    }

    _getStateWithoutCache(habItem, callback) {
        let myURL = this._getURL(`/rest/items/${habItem}/state`).href;
        this._log.debug(`Getting value for ${habItem} from openHAB`);
        needle.get(
            myURL, 
            function (error, response, body) {
                let returnedError = null;
                let returnedValue = null;
                if(error) {
                    returnedError = error;
                } else if (response.statusCode === 404) {
                    returnedError = new Error(`Item does not exist!`);
                } else if (!(body)) {
                    returnedError = new Error(`Unable to retrieve state`);
                } else {
                    returnedValue = this._cleanOpenHABState(body);
                    this._log.warn(`Caching value ${returnedValue} for ${habItem}`);
                    this._valueCache.set(habItem, returnedValue);
                }

                if(callback !== null && callback !== undefined && typeof (callback) === "function") {
                    callback(returnedError, returnedValue);
                }
            }.bind(this)
        )
    }

    getStateSync(habItem) {
        if(this._valueCache.get(habItem)) {
            this._log.debug(`Getting value for ${habItem} from the cache`);
            return this._valueCache.get(habItem);
        } else {
            this._log.warn(`Getting value for ${habItem} from openHAB, because no cached state exists`);
            let myURL = this._getURL(`/rest/items/${habItem}/state`).href;
            const response = syncRequest('GET', myURL);
            if (response.statusCode === 404) {
                return new Error(`Item does not exist!`);
            } else if (!(response.body)) {
                return new Error(`Unable to retrieve state`);
            } else {
                let value = this._cleanOpenHABState(response.body.toString('ASCII'));
                this._valueCache.set(habItem, value);
                return value;
            }
        }
    }

    sendCommand(habItem, command, callback) {
        if(this._valueCache.exists(habItem)) {
            this._log.debug(`Invalidating cache for ${habItem}`);
            this._valueCache.del(habItem);
        }
        let myURL = this._getURL(`/rest/items/${habItem}`).href;
        var plainTextHeader = {
            headers: {
                'Content-Type': 'text/plain'
            }
        }
        needle.post(
            myURL, 
            command,
            plainTextHeader,
            function(error, response) {
                if(error) {
                    callback(error);
                } else if (response.statusCode === 404) {
                    callback(new Error(`Item does not exist!`));
                } else if (response.statusCode === 400) {
                    callback(new Error(`Item command null`));
                } else {
                    callback(null);
                }
            }
        )
    }

    updateState(habItem, state, callback) {
        if(this._valueCache.exists(habItem)) {
            this._log.debug(`Invalidating cache for ${habItem}`);
            this._valueCache.del(habItem);
        }
        let myURL = this._getURL(`/rest/items/${habItem}/state`).href;
        needle.put(myURL, state, function(error, response) {
            if(error) {
                callback(error);
            } else if (response.statusCode === 404) {
                callback(new Error(`Item does not exist!`));
            } else if (response.statusCode === 400) {
                callback(new Error(`Item state null`));
            } else {
                callback(null);
            }
        })
    }

    // Will call callback with callback(error, type)
    getItemType(habItem) {
        return this._typeCache.get(habItem);
    }

    syncItemTypes() {
        this._log.info(`Syncing all items & types from openHAB`);
        let myURL = this._getURL(`/rest/items`, `recursive=false&fields=name%2Ctype%2Ceditable`).href;
        const response = syncRequest('GET', myURL);
        if (response.statusCode !== 200) {
            return new Error(`Unable to get items: HTTP code ${response.statusCode}!`);
        } else {
            const items = JSON.parse(response.body);
            if(items.length > 0) {
                this._log.debug(`Got array with ${items.length} item/s`);
                items.forEach(function(item) {
                    let type = this._cleanOpenHABType(item.type);
                    this._log.debug(`Got item ${item.name} of type ${type}, adding to type cache`);
                    this._typeCache.set(item.name, type);
                }.bind(this));
            } else {
                this._log.error(`Received no items from openHAB, unable to sync states!`);
            }
        }
        this._log.debug(`Done syncing all items & types!`);
    }

    syncItemValues() {
        this._log.info(`Syncing all item values from openHAB`);
        let myURL = this._getURL(`/rest/items`, `recursive=false&fields=name%2Cstate%2Ceditable`).href;
        const response = syncRequest('GET', myURL);
        if (response.statusCode !== 200) {
            return new Error(`Unable to get item values: HTTP code ${response.statusCode}!`);
        } else {
            const items = JSON.parse(response.body);
            if(items.length > 0) {
                this._log.debug(`Got array with ${items.length} item/s`);
                items.forEach(function(item) {
                    if(this._subscriptions[item.name] !== undefined) {
                        let state = this._cleanOpenHABState(item.state);
                        this._log.debug(`Got item ${item.name} with value ${state}, adding to value cache`);
                        this._valueCache.set(item.name, state);
                    } else {
                        this._log.debug(`Got item ${item.name} with value ${item.state}, not adding to value cache, since it is not linked to homebridge!`);
                    }
                }.bind(this));
            } else {
                this._log.error(`Received no items from openHAB, unable to sync states!`);
            }
        }
        this._log.debug(`Done syncing all items values!`)
    }

    getOpenHABAPIVersion() {
        this._log.debug(`Trying to identify openHAB API Version for host (${this._hostname})...`);
        let myURL = this._getURL(`/rest/`);
        const response = syncRequest('GET', myURL);
        if (response.statusCode !== 200) {
            return new Error(`Unable to get openHAB API Version: HTTP code ${response.statusCode}!`);
        } else {
            const parsedBody = JSON.parse(response.body);
            let version = parseInt(parsedBody.version);
            this._log.debug(`openHAB API Version for host (${this._hostname}) is (${version}).`);
            return version;
        }
    }

    getItemsTopic() {
        if (typeof this._apiVersion === 'undefined') {
            this._apiVersion = this.getOpenHABAPIVersion(); 
        }
        if(this._apiVersion >= 4) {
            return 'openhab/items/*';
        }
        else {
            return 'smarthome/items/';
        }
    }

    subscribe(habItem, callback) {
        if(!this._subscriptions[habItem]) {
            this._subscriptions[habItem] = [];
        }
        this._log.debug(`Queueing subscription for ${habItem}`);
        this._subscriptions[habItem].push(callback);
    }

    startSubscription() {
        let myURL = this._getURL('/rest/events',`topics=${this.getItemsTopic()}`).href;
        const CLOSED = 2;

        let source = new EventSource(myURL);
        source.onmessage = function (eventPayload) {
            let eventData = JSON.parse(eventPayload.data);
            if (eventData.type === "ItemStateChangedEvent") {
                let item = eventData.topic.replace(this.getItemsTopic().replace("*", ""), "").replace("/statechanged", "");
                let value = this._cleanOpenHABState(JSON.parse(eventData.payload).value);

                if(this._subscriptions[item] !== undefined) {
                    this._log.debug(`Received new state for item ${item}: ${value}`);
                    this._valueCache.set(item, value);
                    this._subscriptions[item].forEach(function(callback){
                        callback(value, item);
                    });
                } else {
                    this._log.debug(`Ignoring new state for item ${item} (${value}), because it is not registered with homebridge`);
                }
            }
        }.bind(this);
        source.onerror = function (err) {
            if (err.message) {
                let msg;
                if (err.status) {
                    msg = `${err.status}: ${err.message}`;
                } else {
                    msg = err.message;
                }
                if (source.readyState === CLOSED || err.status === 404) {
                    msg = `Subscription service closed, trying to reconnect in 1sec...`;
                    setTimeout(function () {
                        this._log.warn(`Trying to reconnect subscription service...`);
                        source.close();
                        this.startSubscription();
                    }.bind(this), 1000);
                }
                this._log.error(msg);
            }
        }.bind(this);
    }

    // This function is called before a value received from openHAB was passed to the cache or homebridge application
    _cleanOpenHABState(value) {
        // This checks if the value is a number (eventually followed by a unit) and extracts only the number
        let matchedValue =  value.match(/^\d+((\.|,)\d*)*/i);
        if (matchedValue) {
            // This checks if the value is a number followed by a scientific exponent (e.g. `7E+1`), normalizing notation
            let exponentMatch = value.match(/E\+\d*/i);
            if(exponentMatch) {
                let tempValue = matchedValue[0] * Math.pow(10, parseInt(exponentMatch[0].substring(2)));
                this._log.debug(`Recognized number with potential unit and scientific exponent (${value}), normalizing to ${tempValue}`);
                return `${tempValue}`;
            } else {
                this._log.debug(`Recognized number with potential unit (${value}), extracting only the number: ${matchedValue[0]}`);
                return matchedValue[0];
            }
        } else {
            return value;
        }
    }

    // This function is called before a type received from openHAB was passed to the cache or homebridge application
    _cleanOpenHABType(type) {
        if(type.startsWith("Number")) {
            this._log.debug(`Received item type ${type}, transforming to 'Number'`);
            return "Number";
        } else {
            return type;
        }
    }
}

module.exports = {OpenHAB};
